### PR TITLE
Fix KTLS with BIO_new_connect

### PR DESCRIPTION
**Pull request by bernd-edlinger**
When a socket connection is done using BIO_new_connect,
the ktls_enable is done too early, and fails with ENOTCONN.
Therefore the KLTS ioctl will fail later with ENOPROTOOPT.
Fix that by doing the ktls_enable after the connection
succeeded, not when the socket is created as that will
always fail.

One example where this happens is doit_localhost in
test/ssl_old_test.c, and therefore, contrary to the expectation
the -client_ktls option did never enable the client KTLS
connection, but this was not noticed, because there was no
diagnostic output, and it was only visible with strace output.

Also enhanced the ssl_old_test -client_ktls/-server_ktls
options together with -v option to print a summary line
if and how KTLS was negotiated in server and client.

While I am already there adjusted the usage info of
the -s_cert, -s_key commands, and allow -time to print the
timings of ktls connections.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: crypto/bio/bss_conn.c [hunk 1]
```diff
@@ -208,8 +208,18 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
...
+            } else {
                 c->state = BIO_CONN_S_OK;
+# ifndef OPENSSL_NO_KTLS
+                /*
+                 * The new socket is created successfully regardless of ktls_enable.
+                 * ktls_enable doesn't change any functionality of the socket, except
+                 * changing the setsockopt to enable the processing of ktls_start.
+                 * Thus, it is not a problem to call it for non-TLS sockets.
+                 */
+                ktls_enable(b->num);
```
Comment (bernd-edlinger) [inline crypto/bio/bss_conn.c:220]: Hmm, in my tests this call retuns sometimes ENOTCONN.
So it's an improvement, but does not 100% work for nonblocking connects,
and that's what doit_localhost does.
The man page says
```
EINPROGRESS
              The socket is nonblocking and the connection cannot be completed
              immediately.   (UNIX domain sockets failed with EAGAIN instead.)
              It is possible to select(2) or poll(2) for completion by select‐
              ing the socket for writing.  After select(2) indicates writabil‐
              ity, use getsockopt(2) to read  the  SO_ERROR  option  at  level
              SOL_SOCKET to determine whether connect() completed successfully
              (SO_ERROR is zero) or unsuccessfully (SO_ERROR  is  one  of  the
              usual  error  codes  listed  here, explaining the reason for the
              failure).
```
we are not doing the select here, probably that's the problem,
I don't see how the bss_conn.c object can signal a SSL_ERROR_WANT_WRITE...

**Comment by bernd-edlinger** [comment]
BTW, in general I see only 50-90% data throughput when KTLS is activated, in comparison to when KTLS is not activated.
I use this:
`CTLOG_FILE=test/ct/log_list.cnf ./util/shlib_wrap.sh ./test/ssl_old_test -s_cert apps/server.pem -ipv6 -server_ktls -client_ktls -bytes 1000000000 -v -time -f`
to measure numbers with KTLS vs. without KTLS.

**Comment by tom-cosgrove-arm** [comment]
Is it okay to fix the spelling of "KTLS" in the title of this PR, or will that cause problems for something?

**Comment by t8m** [comment]
It would be also nice to adjust the commit message @bernd-edlinger.

**Review by t8m — APPROVED** [review]

File: crypto/bio/bio_sock.c [hunk 2]
```diff
@@ -400,7 +400,7 @@ int BIO_socket_wait(int fd, int for_read, time_t max_time)
         return 1;
 
     now = time(NULL);
-    if (max_time <= now)
+    if (max_time < now)
```
Comment (mattcaswell) [inline crypto/bio/bio_sock.c:403]: Why is this change necessary?

File: crypto/bio/bss_conn.c [hunk 3]
```diff
@@ -191,6 +191,9 @@ static int conn_state(BIO *b, BIO_CONNECT *c)
             break;
 
         case BIO_CONN_S_BLOCKED_CONNECT:
+            /* wait for socket being writable, before querying BIO_sock_error */
+            if (BIO_socket_wait(b->num, 0, time(NULL)) == 0)
+                break;
```
Comment (mattcaswell) [inline crypto/bio/bss_conn.c:196]: Why do we need to do this?

File: test/ssl_old_test.c [hunk 4]
```diff
@@ -1296,7 +1296,7 @@ int main(int argc, char *argv[])
     }
 
     if (print_time) {
-        if (bio_type != BIO_PAIR) {
+        if (bio_type == BIO_MEM) {
```
Comment (mattcaswell) [inline test/ssl_old_test.c:1299]: Why this change?

File: test/ssl_old_test.c [hunk 5]
```diff
@@ -2144,8 +2144,25 @@ int doit_localhost(SSL *s_ssl, SSL *c_ssl, int family, long count,
...
+            BIO_printf(bio_stdout, "Server using Kernel TLS for receiving\n");
+
+        if (BIO_get_ktls_send(SSL_get_wbio(c_ssl))
+                && BIO_get_ktls_recv(SSL_get_rbio(c_ssl)))
+            BIO_printf(bio_stdout, "Client using Kernel TLS in both directions\n");
+        else if (BIO_get_ktls_send(SSL_get_wbio(c_ssl)))
+            BIO_printf(bio_stdout, "Client using Kernel TLS for sending\n");
+        else if (BIO_get_ktls_recv(SSL_get_rbio(c_ssl)))
+            BIO_printf(bio_stdout, "Client using Kernel TLS for receiving\n");
+    }
```
Comment (mattcaswell) [inline test/ssl_old_test.c:2165]: The additional info logging would perhaps have been better in a different commit. But never mind.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by bernd-edlinger** [comment]
okay, you mean s/ktls/KTLS/ ?

File: crypto/bio/bio_sock.c (hunk 2 above)
Comment (bernd-edlinger) [reply crypto/bio/bio_sock.c:403]: Yes, I need something to get writability of a socket without waiting the full second.
this just allows the return value to be 1 in this case, I consider it a bug that this
function did always return 0 when the due time is reached, even if the socket
is writable now.

File: crypto/bio/bss_conn.c (hunk 3 above)
Comment (bernd-edlinger) [reply crypto/bio/bss_conn.c:196]: Apparently the connect is not yet completed when BIO_sock_error returns 0
here, and the ktls_enable call fails with ENOTCONN.  That is a race condition,
and does not happen every time.

The man page says

```
EINPROGRESS
              The socket is nonblocking and the connection cannot be completed
              immediately.   (UNIX domain sockets failed with EAGAIN instead.)
              It is possible to select(2) or poll(2) for completion by select‐
              ing the socket for writing.  After select(2) indicates writabil‐
              ity, use getsockopt(2) to read  the  SO_ERROR  option  at  level
              SOL_SOCKET to determine whether connect() completed successfully
              (SO_ERROR is zero) or unsuccessfully (SO_ERROR  is  one  of  the
              usual  error  codes  listed  here, explaining the reason for the
              failure).
```

But we are not doing the select here, probably that's the problem.
So this avoids the ktls_enable from failing, and makes the test case
use the client ktls successfully (modulo the possible #18276 issue).

File: crypto/bio/bio_sock.c (hunk 2 above)
Comment (mattcaswell) [reply crypto/bio/bio_sock.c:403]: Ah - right. Makes sense.

File: test/ssl_old_test.c (hunk 4 above)
Comment (bernd-edlinger) [reply test/ssl_old_test.c:1299]: I want to measure the time using -ipv4/6 -time
and this overrides the bio_type == BIO_IPV4, BIO_IPV6
which are clones of BIO_PAIR and are able to do the timimgs,
in contrast to BIO_MEM, which is not doing any timimgs.

Comment (bernd-edlinger) [reply test/ssl_old_test.c:1299]: But BIO_PAIR is using memory bio, and cannot use KTLS.

File: crypto/bio/bss_conn.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/bio/bss_conn.c:196]: Ok

**Comment by bernd-edlinger** [comment]
I've changed the commit message.  No other changes this time.

File: test/ssl_old_test.c (hunk 4 above)
Comment (mattcaswell) [reply test/ssl_old_test.c:1299]: Ok

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by bernd-edlinger** [comment]
Pushed to master as 598bd77 and 3.0 as c6c3602.
Thanks!

**Comment by bernd-edlinger** [comment]
The problem is that the SSL_set_bio is racing with the socket connect state, and the ktls_enable is failing
when the socket is non-blocking.

**Comment by bernd-edlinger** [comment]
no, I just used select with zero time-out.  If the select fails I can retry the operation. That can't be done in SSL_set_fd,
if that is called with a socket a millisecond to early, it just wont enable KTLS at all.
Are you saying that your driver cant enable the KTLS socket once it is connected?

**Comment by bernd-edlinger** [comment]
@bsdjhb can you please clarify what happens in your enviroment when ktls_enable is called
immediately after the socket enters the connect state (and not immediately after it is created)?

Also note that in the system test the sockets are created here: 
https://github.com/openssl/openssl/blob/8b97bfcccc4328c65156bff6886db8733df39fde/test/helpers/ssltestlib.c#L789-L821

So the ktls_enable happens after the connect, and the connect is done while the socket is still
in blocking mode. If you require the ktls_enable before the connect, that test is likely broken.
